File: src/policies/guild-policy.ts
export interface PolicyCheckResult {
  allowed: boolean;
  reason?: string;
}

export class GuildPolicyEngine {
  static checkActionAllowed(
    actionType: string,
    settings?: {
      enabled?: boolean;
      allowedActions?: string[];
      disabledActions?: string[];
    } | null
  ): PolicyCheckResult {
    if (!settings) return { allowed: true };

    if (settings.enabled === false) {
      return { allowed: false, reason: 'Bot interactions are currently disabled for this server.' };
    }

    if (settings.disabledActions && settings.disabledActions.includes(actionType)) {
      return { allowed: false, reason: `Action '${actionType}' has been disabled by server administrators.` };
    }

    if (
      settings.allowedActions &&
      settings.allowedActions.length > 0 &&
      !settings.allowedActions.includes(actionType)
    ) {
      return { allowed: false, reason: `Action '${actionType}' is not in the server's allowed actions list.` };
    }

    return { allowed: true };
  }
}

File: bot-py/src/policies/guild_policy.py
from typing import Any


class GuildPolicyEngine:
    @staticmethod
    def check_action_allowed(
        action_type: str,
        settings: Any | None = None
    ) -> tuple[bool, str | None]:
        if not settings:
            return True, None

        if getattr(settings, "enabled", True) is False:
            return False, "Bot interactions are currently disabled for this server."

        disabled_actions: list[str] = getattr(settings, "disabledActions", []) or []
        if action_type in disabled_actions:
            return False, f"Action '{action_type}' has been disabled by server administrators."

        allowed_actions: list[str] = getattr(settings, "allowedActions", []) or []
        if allowed_actions and action_type not in allowed_actions:
            return False, f"Action '{action_type}' is not in the server's allowed actions list."

        return True, None
